src/Data/Nat.agda

```diff
@@ -28,6 +38,35 @@ open import Data.Nat.Properties public
   ; _<‴?_; _≤‴?_; _≥‴?_; _>‴?_
   )
 
+------------------------------------------------------------------------
+-- Reading a number
+
+readMaybe : ∀ base {base≤16 : True (base ≤? 16)} → String → Maybe ℕ
+readMaybe _ "" = nothing
+readMaybe base = Maybe.map convert
+              ∘′ TraversableA.mapA Maybe.applicative readDigit
+              ∘′ String.toList
+
+  where
+
+    convert : List ℕ → ℕ
+    convert = List.foldl (λ acc d → base * acc + d) 0
```

Comment:
I was wondering if we could make use of some of the machinery from `Data.Digit` to help us, but it looks like that goes via `Fin` rather than `ℕ` so maybe not...

Reply:
Yeah I had started working in `Data.Digit` but gave up because of `Fin`.
We may want a `read` there too but the one for `Nat` should definitely
be implemented efficiently IMO.